internal/wait/wait.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the Apache License, Version 2.0, see LICENCE file for details.
+
+package wait
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/juju/clock"
+	"github.com/juju/retry"
+	"github.com/juju/terraform-provider-juju/internal/juju"
+)
+
+type Logger interface {
+	Debugf(msg string, additionalFields ...map[string]interface{})
+}
+
+// RetryConf is a struct to configure the retry behavior.
+type RetryConf struct {
+	// MaxDuration is the maximum duration to wait for the condition to be met.
+	MaxDuration time.Duration
+	// Delay is the delay between retries.
+	Delay time.Duration
+	// Clock is the clock to use for timing.
+	Clock clock.Clock
+}
+
+// WaitForCfg is a configuration structure for the WaitFor function.
+type WaitForCfg[I any, D any] struct {
+	Context context.Context
+
+	// GetData is a function that retrieves data based on the input.
+	GetData GetData[I, D]
+	// Input is the input to be passed to the GetData function.
+	Input I
+	// DataAssertions is a list of assertions to check the data against.
+	// If any assertion fails, the function will return an error.
+	DataAssertions []Assert[D]
+	// NonFatalErrors is a list of non-fatal errors to ignore.
+	NonFatalErrors []error
+
+	// RetryConf is a configuration for retrying the operation.
+	// If not provided, default values will be used.
+	RetryConf *RetryConf
+}
+
+// WaitForErrorCfg is a configuration structure for the WaitForError function.
+type WaitForErrorCfg[I any, D any] struct {
+	Context context.Context
+
+	// GetData is a function that retrieves data based on the input.
+	GetData GetData[I, D]
+	// Input is the input to be passed to the GetData function.
+	Input I
+	// ErrorToWait is the error to wait for.
+	ErrorToWait error
+	// NonFatalErrors is a list of non-fatal errors to ignore.
+	NonFatalErrors []error
+
+	// RetryConf is a configuration for retrying the operation.
+	// If not provided, default values will be used.
+	RetryConf *RetryConf
+}
+
+type GetData[I any, D any] func(I) (D, error)
```

Comment:
godoc, please.. also, we might want to pass `context.Context` into the GetData function